crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -180,7 +197,31 @@ impl Memory {
     #[inline(always)]
     #[cfg_attr(debug_assertions, track_caller)]
     pub fn copy(&mut self, dst: usize, src: usize, len: usize) {
-        self.data.copy_within(src..src + len, dst);
+        self.context_memory_mut().copy_within(src..src + len, dst);
+    }
+
+    /// Get a reference to the memory of the current context
+    #[inline(always)]
+    fn context_memory(&self) -> &[u8] {
+        let last_checkpoint = self.last_checkpoint();
+        let current_len = self.current_len;
+        // Safety: it is a valid pointer to a slice of `self.data`
+        &self.data[last_checkpoint..last_checkpoint + current_len]
+    }
+
+    /// Get a mutable reference to the memory of the current context
+    #[inline(always)]
+    fn context_memory_mut(&mut self) -> &mut [u8] {
+        let last_checkpoint = self.last_checkpoint();
+        let current_len = self.current_len;
+        // Safety: it is a valid pointer to a slice of `self.data`
+        &mut self.data[last_checkpoint..last_checkpoint + current_len]
+    }
+
+    /// Get the last memory checkpoint
+    #[inline(always)]
+    fn last_checkpoint(&self) -> usize {
+        *self.checkpoints.last().unwrap_or(&0)
```

Comment:
```suggestion
        self.checkpoints.last().cloned().unwrap_or_default()
```
`cloned` would remove the reference from the option.